### PR TITLE
fix(websocket-websys): change close code in drop implementation to 1000

### PR DESCRIPTION
**Pull request by sisou**
## Description

In browsers, only the code 1000 and codes between 3000-4999 are allowed to be set by userland code: https://websockets.spec.whatwg.org/#dom-websocket-close

I found this while debugging use-after-free errors in our WASM application. Turns out, when connections are timing out, libp2p in rust drops them, but does not seem to close them explicitly (we are using libp2p-swarm). This led to these WebSockets still emitting events, the handlers of which were already dropped on the rust side, though.

A long investigation led me to have a look into the `Result` that gets returned from `close_with_code_and_reason`, and it turns out it's an error! Specifically:
```
InvalidAccessError: Failed to execute 'close' on 'WebSocket': The code must be either 1000, or between 3000 and 4999. 1001 is neither.
```

This PR only fixes the failing closing of the WebSocket, not the remaining use-after-free problem.

### Not ignoring error results

Can `Err` results not be ignored, but instead logged? That would have shown this problem a long time ago.

### Use-after-free event handlers

Calling `.close()` on the WebSocket leads to the WebSocket still emitting its `"close"` event (and also an `"error"` event, if the connection is still pending and ran into the time-out), for which the `websocket-websys`  transport registers a handler. This handler is dropped though together with the connection, so when the event comes into WASM, the handler is already free'd and this error is thrown (by wasm-bindgen?):
```
Error: closure invoked recursively or after being dropped
```

This error is not critical in browsers (only annoying in the console) but exits NodeJS programs due to an uncaught exception.

A quick-fix would be to unregister the relevant event handlers in the `Drop` implementation, like so:
```rs
self.inner.socket.set_onclose(None);
self.inner.socket.set_onerror(None);
```

According to @nibhar, there must be a better way, though. He suggests storing the socket's event handlers outside the connection, having them clean-up themselves when the onclose handler is called.

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by sisou** [comment]
Here is a basic reproduction which I used to create this PR: https://github.com/sisou/libp2p-websocket-reproduction

**Review by jxs — CHANGES_REQUESTED** [review]
Hi, and thanks for the thorough report! Looks good to me, can you also add an entry in the `CHANGELOG.md`?
 cc @thomaseizinger you were the one adding the code, you may want to review it, and probably have an opinion on the use after free situation.

**Review by thomaseizinger — APPROVED** [review]
Wow, that is surprising!

Great find. Yeah needs a version bump and changelog entry.

**Comment by jsdanielh** [comment]
> Wow, that is surprising!
> 
> Great find. Yeah needs a version bump and changelog entry.

Thanks for reviewing it. We just added the version bump and changelog entry.

File: transports/websocket-websys/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,10 @@
+## 0.3.2
+
+- Change close code in drop implementation to `1000` given than in browsers only
```
Comment (jxs) [inline transports/websocket-websys/CHANGELOG.md]:
```suggestion
- Change close code in drop implementation to `1000` given that in browsers only
```

**Review by jxs — APPROVED** [review]

**Comment by jxs** [comment]
Thanks! Just waiting on https://github.com/libp2p/rust-libp2p/pull/5253 so we can get CI green again

**Review by jxs — APPROVED** [review]

**Comment by jxs** [comment]
@jsdanielh do you require a release?

**Comment by jsdanielh** [comment]
> @jsdanielh do you require a release?

Yeah, that would be great

**Comment by jxs** [comment]
sorry for the delay @jsdanielh and thanks for your patience.
A `libp2p-websocket-websys` version has now been released, see https://github.com/libp2p/rust-libp2p/discussions/5351 for more information.

**Pull request merged**